### PR TITLE
Clean up UART

### PR DESCRIPTION
**Pull request by bugadani**
- Drop unnecessary constructors
- Drop unnecessary default pin types
  - These aren't very useful as we need to know their numbers anyway
- Reduce the number of pin connection implementations, in preparation for #2128

### CODE REVIEW

File: esp-hal/src/uart.rs [hunk 1]
```diff
@@ -2136,14 +2064,10 @@ mod asynch {
...
-            crate::into_ref!(rx);
-            crate::into_ref!(tx);
-            tx.set_to_push_pull_output(Internal);
-            tx.connect_peripheral_to_output(T::tx_signal(), Internal);
-
-            rx.init_input(false, false, Internal);
-            rx.connect_input_to_peripheral(T::rx_signal(), Internal);
-            let mut this = Self::new_with_config_inner(uart, config)?;
+            let mut this = UartBuilder::new(uart)
+                .with_tx(tx)
```
[the remark this answers is not in the document]
Comment (MabezDev) [reply esp-hal/src/uart.rs:2068]: Possibly unrelated, but https://github.com/esp-rs/esp-hal/blob/7332f5834a42254906edaecf6af8adfbe0370fbb/esp-hal/src/gpio/mod.rs#L744 isn't required as per the TRM, you can read the input state regardless of the selected signal, also it probably should have been `func_in_sel_cfg`.

Other than that observation, I'm not sure why the ordering would ever matter here :thinking:

File: hil-test/tests/uart_async.rs [hunk 2]
```diff
@@ -7,7 +7,7 @@
 //! Connect TX and RX pins.
 
 //% CHIPS: esp32 esp32c2 esp32c3 esp32c6 esp32h2 esp32s2 esp32s3
-//% FEATURES: generic-queue
```
Comment (Dominaezzz) [inline hil-test/tests/uart_async.rs:10]: Was this intentional?

Comment (bugadani) [reply hil-test/tests/uart_async.rs]: Doing it, yes, leaving it in, no :D thanks

**Comment by bjoernQ** [comment]
I guess people could figure it out on their own but should we mention how to migrate code using `*_with_default_pins` in the migration guide?

**Review by bjoernQ — APPROVED** [review]
Looks good to me

**Review by MabezDev — APPROVED** [review]
Thanks!

**Pull request merged**